pkg/scope/provider.go

```diff
@@ -106,6 +128,34 @@ func NewProviderScope(cloud clientconfig.Cloud, caCert []byte, logger logr.Logge
 	}, nil
 }
 
+func NewCachedProviderScope(cache *cache.LRUExpireCache, cloud clientconfig.Cloud, caCert []byte, logger logr.Logger) (Scope, error) {
+	key, err := getScopeCacheKey(cloud)
+	if err != nil {
+		return nil, fmt.Errorf("can't compute cloud config cache key: %w", err)
+	}
+
+	if scope, found := cache.Get(key); found {
+		logger.V(6).Info("Using scope from cache")
+		return scope.(Scope), nil
+	}
+
+	scope, err := NewProviderScope(cloud, caCert, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := scope.ExtractToken()
+	if err != nil {
+		return nil, err
+	}
+
+	// compute the token expiration time
+	expiry := time.Until(token.ExpiresAt) / 2
```

Comment:
Out of curiosity:
Is using half expiry time good for time out of sync issues or what is the reason for choosing `/ 2` here?